spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java

```diff
@@ -0,0 +1,147 @@
+package edu.umd.cs.findbugs.detect;
+
+import org.apache.bcel.Const;
+
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.StringAnnotation;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
+
+public class DateFormatStringChecker extends OpcodeStackDetector {
+
+    /** List of bytecode instructions which could signal a method we should check. */
+    private static final List<Integer> CONST_ARRAY_LIST = new ArrayList<>(Arrays.asList((int) Const.INVOKESPECIAL,
+            (int) Const.INVOKEVIRTUAL, (int) Const.INVOKESTATIC, (int) Const.INVOKEINTERFACE));
+
+    /** List of operands matching target pattern. */
+    private static final String SIG_CONSTANT_OPERAND_1 = "Ljava/lang/String;)V";
+    private static final String SIG_CONSTANT_OPERAND_2 = "Ljava/lang/String;Ljava/util/Locale;)V";
+    private static final String SIG_CONSTANT_OPERAND_3 = "Ljava/lang/String;Ljava/text/DateFormatSymbols;)V";
+
+    private static final String CLASS_CONSTANT_OPERAND = "java/text/SimpleDateFormat";
+
+    private static final String NAME_CONSTANT_OPERAND_1 = "<init>";
+    private static final String NAME_CONSTANT_OPERAND_2 = "applyPattern";
+    private static final String NAME_CONSTANT_OPERAND_3 = "applyLocalizedPattern";
+
+    /** Creates DateFormatRule objects for different bad combinations of date format flags. */
+    private static final DateFormatStringChecker.DateFormatRule STANDARD_TIME_FLAG_1 =
+            new DateFormatStringChecker.DateFormatRule("a", Arrays.asList("h"));
+    private static final DateFormatStringChecker.DateFormatRule STANDARD_TIME_FLAG_2 =
+            new DateFormatStringChecker.DateFormatRule("a", Arrays.asList("K"));
+    private static final DateFormatStringChecker.DateFormatRule MILITARY_FLAG_1 =
+            new DateFormatStringChecker.DateFormatRule(null, Arrays.asList("H", "a"));
+    private static final DateFormatStringChecker.DateFormatRule MILITARY_FLAG_2 =
+            new DateFormatStringChecker.DateFormatRule(null, Arrays.asList("k", "a"));
+    private static final DateFormatStringChecker.DateFormatRule WEEK_YEAR_FLAG =
+            new DateFormatStringChecker.DateFormatRule("w", Arrays.asList("Y", "M", "d"));
+
+    private static final String BUG_TYPE = "FS_BAD_DATE_FORMAT_FLAG_COMBO";
```

Comment:
This field and several other fields are only used by one method. It's better to use local variables instead.